calico/felix/labels.py

```diff
@@ -0,0 +1,480 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Metaswitch Networks
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+
+from calico.calcollections import MultiDict
+from calico.felix.selectors import LabelToLiteralEquality, LabelInSetLiteral
+
+_log = logging.getLogger(__name__)
+
+
+class LinearScanLabelIndex(object):
+    """
+    A label index matches a set of SelectorExpressions against a set of
+    label dicts.  As the matches between the two collections change,
+    it triggers calls to on_match_started/on_match_stopped.
+
+    A simple implementation of a label index.  Every update is handled by
+    a full linear scan.
+
+    This class has two purposes:
+
+    - it provides a benchmark against which other implementations can be
+      measured
+    - since it's simple, it's useful for comparative testing; any
+      other label index implementation should give the same end result.
```

Comment:
I'm finding this file a bit confusing, and I think it's the combination of several factors.

Firstly, that all these classes are called Index, when they should be Map or DB or something.  In a book, the Index is just the bit at the end; these classes contain the whole data and so represent the book, not the index.

Or is it perhaps a bit deeper than that, in that it's not completely clear to me what these classes are doing.  Perhaps they're hard to name clearly because their role is unclear.  They have event-driven inputs and outputs, but these aren't presented anywhere in an overall way.  It seems like it would be a bit of a maze for someone writing new code that _should_ call these classes, to work and when and how to do that.  In particular it's hard to see how LabelInheritanceIndex composes with LabelValueIndex - e.g. do calls to LabelInheritanceIndex's inputs result in calls of LabelValueIndex's outputs?

Also I wonder if the use of inheritance is correct.  From LinearScanLinearIndex to LabelValueIndex, what changes (IIUC) is not the basic data at all, but the method that is used to scan when something in that data changes, and to emit updates.  I wonder if that should be one class to hold the data, with a has-a relationship to some kind of Scanner class?

So what's the upshot of all this?  I'm not sure, but perhaps:

- Look at making the naming more precise and intuitive.  The best name I can think of for LinearScanLinearIndex and LabelValueIndex is actually SelectorMonitor; it reports as selectors become true or false for each of the provided label dicts.

- Provide upfront doc of the inputs and outputs.

- Consider the has-a idea for the Scanner code - I'm afraid I can't see in detail how this might work, so it might be a rubbish idea - but perhaps you can.